### PR TITLE
crosscluster/logical: spruce up ldr debugging vtable

### PR DESCRIPTION
**Pull request by msbutler**
Epic: none

Release note: none

### CODE REVIEW

**Comment by msbutler** [comment]
I plan to fix the zip tests once y'all provide any feedback on the new logs and vtable

**Comment by msbutler** [comment]
Here it is in action:
```
2024/09/24 18:54:55 roachprod.go:546: node 3:
       stream_id      | consumer |        state        |    recv_time    | last_recv_time  |   ingest_time   |   flush_time    | flush_count | flush_kvs | flush_bytes | flush_batches | last_flush_time | last_kvs_done | last_kvs_todo | last_batches |  last_slowest   |  resolved_age   | last_checkpoint | checkpoints | retry_size
----------------------+----------+---------------------+-----------------+-----------------+-----------------+-----------------+-------------+-----------+-------------+---------------+-----------------+---------------+---------------+--------------+-----------------+-----------------+-----------------+-------------+-------------
  1006414674038685697 | 3[0]     | receiving           | 00:04:48.94915  | 00:00:00.63267  | 00:00:21.404689 | 00:00:21.401628 |         124 |    110899 |   121119187 |        110899 | 00:00:00.136364 |           961 |           961 |          961 | 00:00:00.003557 | 00:00:15.714296 | 00:00:10.964188 |          21 |          0
  1006414674038685697 | 3[1]     | processing-flushing | 00:04:55.733003 | 00:00:04.75728  | 00:00:14.620854 | 00:00:14.617705 |          83 |     72692 |    79435155 |         72692 | 00:00:00.050363 |           960 |           388 |          388 | 00:00:00.003921 | 00:00:15.714296 | 00:00:10.974239 |          21 |          0
  1006414674038685697 | 3[2]     | processing-flushing | 00:04:48.806749 | 00:00:02.44678  | 00:00:21.547118 | 00:00:21.544257 |         122 |    108182 |   118135992 |        108182 | 00:00:00.18504  |           961 |           935 |          935 | 00:00:00.016853 | 00:00:15.714296 | 00:00:11.105438 |          21 |          0
  1006414674038685697 | 3[3]     | receiving           | 00:04:53.550198 | 00:00:03.108661 | 00:00:16.803793 | 00:00:16.801106 |         101 |     84967 |    92815960 |         84967 | 00:00:00.165224 |           960 |           960 |          960 | 00:00:00.007025 | 00:00:15.714296 | 00:00:10.334162 |          21 |          0
  1006414674038685697 | 3[4]     | receiving           | 00:04:50.46181  | 00:00:01.619884 | 00:00:19.89214  | 00:00:19.889081 |         117 |    100824 |   110100960 |        100824 | 00:00:00.172656 |           961 |           961 |          961 | 00:00:00.009764 | 00:00:15.714296 | 00:00:10.300963 |          21 |          0
  1006414674038685697 | 3[5]     | receiving           | 00:04:59.063658 | 00:00:04.441944 | 00:00:11.290259 | 00:00:11.288277 |          64 |     52497 |    57341189 |         52497 | 00:00:00.148035 |           961 |           961 |          961 | 00:00:00.004414 | 00:00:15.714296 | 00:00:10.293617 |          21 |          0
  1006414674038685697 | 3[6]     | receiving           | 00:04:42.877264 | 00:00:00.059237 | 00:00:27.476416 | 00:00:27.473373 |         164 |    144367 |   157707717 |        144367 | 00:00:00.195248 |           689 |           689 |          689 | 00:00:00.011014 | 00:00:05.714296 | 00:00:00.059237 |          22 |          0
  1006414674038685697 | 3[7]     | receiving           | 00:04:48.353243 | 00:00:00.02798  | 00:00:22.000486 | 00:00:21.99764  |         121 |    101234 |   110555890 |        101234 | 00:00:00.206579 |           106 |           106 |          106 | 00:00:00.013217 | 00:00:05.714296 | 00:00:00.027981 |          22 |          0
```

File: pkg/ccl/crosscluster/logical/purgatory.go [hunk 1]
```diff
@@ -152,7 +156,7 @@ func (p *purgatory) Drain(ctx context.Context) error {
 	return nil
 }
 
-func (p purgatory) Empty() bool {
+func (p *purgatory) Empty() bool {
```
Comment (dt) [inline pkg/ccl/crosscluster/logical/purgatory.go:159]: eh, why?

File: pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go [hunk 2]
```diff
@@ -216,7 +216,7 @@ func newLogicalReplicationWriterProcessor(
 		checkpointCh:   make(chan []jobspb.ResolvedSpan),
 		errCh:          make(chan error, 1),
 		logBufferEvery: log.Every(30 * time.Second),
-		debug: streampb.DebugLogicalConsumerStatus{
+		debug: &streampb.DebugLogicalConsumerStatus{
```
Comment (dt) [inline pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go:219]: Why the extra alloc?

File: pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go [hunk 3]
```diff
@@ -231,13 +231,16 @@ func newLogicalReplicationWriterProcessor(
...
 		bytesGauge:  lrw.metrics.RetryQueueBytes,
 		eventsGauge: lrw.metrics.RetryQueueEvents,
+		debug:       lrw.debug,
 	}
 
 	if err := lrw.Init(ctx, lrw, post, logicalReplicationWriterResultType, flowCtx, processorID, nil, /* memMonitor */
 		execinfra.ProcStateOpts{
 			InputsToDrain: []execinfra.RowSource{},
 			TrailingMetaCallback: func() []execinfrapb.ProducerMetadata {
+				log.Infof(lrw.Ctx(), "logical replication writer processor closing")
```
Comment (dt) [inline pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go:241]: I think we'd be fine with just one of these two lines?

File: pkg/repstream/streampb/empty.go [hunk 4]
```diff
@@ -154,62 +175,86 @@ func (d *DebugLogicalConsumerStatus) GetStats() DebugLogicalConsumerStats {
...
+	d.mu.stats.Flushes.Last.Batches = 0
+	d.mu.stats.Flushes.Last.ProcessedKVs = 0
+	d.mu.stats.Flushes.Last.LastFlushNanos = 0
+
+	d.mu.stats.CurrentState = FlushingState
+	d.mu.stats.Flushes.Last.CurFlushStart = start
+	d.mu.stats.Flushes.Last.TotalKVs = keyCount
 	failPercent := d.mu.injectFailurePercent
-	d.mu.Unlock()
+	d.mu.Unlock() // nolint:deferunlockcheck
```
Comment (dt) [inline pkg/repstream/streampb/empty.go:220]: I'd just defer it at this point; too many lines away to be relying on inspection of every line now.

File: pkg/repstream/streampb/empty.go [hunk 5]
```diff
@@ -129,23 +130,43 @@ type DebugLogicalConsumerStatus struct {
 	}
 }
 
+const (
+	ProcessingState = "processing-other"
```
Comment (dt) [inline pkg/repstream/streampb/empty.go:134]: These look wide; think about the table we're trying to consume and how hard it is to read if it wraps to multiple lines per row. I'd do `wait`, `flush`, and if you really want a non-flush other state, `other`

Comment (dt) [reply pkg/repstream/streampb/empty.go:134]: I'd also move these strings to crdb_internal; the only thing we need in this struct is a iota byte

**Comment by dt** [comment]
I'd move resolved_age all the way to the right again; it's the most important column so being able to just scan the edge is nice.

File: pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go (hunk 3 above)
Comment (msbutler) [reply pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go]: Removed the second one.

File: pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go (hunk 2 above)
Comment (msbutler) [reply pkg/ccl/crosscluster/logical/logical_replication_writer_processor.go]: oops. fixed.

File: pkg/repstream/streampb/empty.go (hunk 5 above)
Comment (msbutler) [reply pkg/repstream/streampb/empty.go]: shortened the names. Created a string method in streampb to be called in crdb_internal.

File: pkg/ccl/crosscluster/logical/purgatory.go (hunk 1 above)
Comment (msbutler) [reply pkg/ccl/crosscluster/logical/purgatory.go]: hrm, the compiler was complaining at some point about the mixed receiver methods, but no longer. Fixed.

File: pkg/repstream/streampb/empty.go (hunk 4 above)
Comment (msbutler) [reply pkg/repstream/streampb/empty.go]: Done. I added a defer some other other chunky functions as well.

**Review by msbutler — COMMENTED** [review]
Done, and moved resolved_age all the way to the right.

**Review by dt — APPROVED** [review]

**Comment by msbutler** [comment]
unrelated flake

TFTR

bors r=dt

**Pull request merged**